server/src/main/java/org/elasticsearch/index/mapper/ParsedDocument.java

```diff
@@ -35,6 +38,37 @@
 
     private Mapping dynamicMappingsUpdate;
 
+    public static ParsedDocument noopTombstone(String index, String reason) {
+        ParsedDocument doc = deleteTombstone(index, null);
+        // Store the reason of a noop as a raw string in the _source field
+        final BytesRef byteRef = new BytesRef(reason);
+        doc.rootDoc().add(new StoredField(SourceFieldMapper.NAME, byteRef.bytes, byteRef.offset, byteRef.length));
+        return doc;
+    }
+
+    public static ParsedDocument deleteTombstone(String index, String id) {
+        SourceToParse source = new SourceToParse(index, id == null ? "" : id, new BytesArray("{}"), XContentType.JSON);
+        ParseContext.InternalParseContext context = new ParseContext.InternalParseContext(null, 1,null, source, null);
```

Comment:
I am a bit nervous about these null arguments. I see why you are using them, to avoid carrying around stuff that is not needed. On the other hand, it is arguments that I would not normally expect to be null within the ParseContext and they are nullable specifically in the context of the tombstone usages, which I would consider an edge case. Would it make sense to change the additional constructor that you added to not take those arguments, enforce non-null values in the other constructor? Would it even make sense to enforce that you don't retrieve what is null through assertions?

Reply:
I've pulled most of this out into a private `tombstone` method so we don't start throwing nulls around.

Reply:
but the context still takes null values for three arguments which are normally not nullable, I think?